src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxTransformNode.cs

```diff
@@ -0,0 +1,118 @@
+
+        private sealed class Builder : ISyntaxTransformBuilder
+        {
+            private readonly SyntaxTransformNode<T> _owner;
+
+            private readonly NodeStateTable<T> _previousTransformTable;
+            private readonly NodeStateTable<T>.Builder _transformTable;
+
+            private readonly NodeStateTable<SyntaxNode> _previousFilterTable;
+            private readonly NodeStateTable<SyntaxNode>.Builder _filterTable;
+
+            public Builder(SyntaxTransformNode<T> owner, NodeStateTable<T> previousTransform, NodeStateTable<SyntaxNode> previousFilter)
+            {
+                _owner = owner;
+
+                _previousFilterTable = previousFilter;
+                _filterTable = new NodeStateTable<SyntaxNode>.Builder();
+
+                _previousTransformTable = previousTransform;
+                _transformTable = new NodeStateTable<T>.Builder();
+            }
+
+            public bool Filter(SyntaxNode syntaxNode) => _owner._filterFunc(syntaxNode);
+
+            public void AddFilterFromPreviousTable(SemanticModel model, EntryState state)
+            {
+                var nodes = _filterTable.AddEntriesFromPreviousTable(_previousFilterTable, state);
+                UpdateTransformTable(nodes, model, state);
+            }
+
+            public void AddFilterEntries(ImmutableArray<SyntaxNode> nodes, SemanticModel model)
+            {
+                _filterTable.AddEntries(nodes, EntryState.Added);
+                UpdateTransformTable(nodes, model, EntryState.Added);
+            }
+
+            private void UpdateTransformTable(ImmutableArray<SyntaxNode> nodes, SemanticModel model, EntryState state)
+            {
+                foreach (var node in nodes)
+                {
+                    if (state == EntryState.Removed)
+                    {
+                        _transformTable.AddEntriesFromPreviousTable(_previousTransformTable, EntryState.Removed);
+                    }
+                    else
+                    {
+                        var value = new GeneratorSyntaxContext(node, model);
+                        var transformed = ImmutableArray.Create(_owner._func(value));
+                        if (_previousTransformTable.IsEmpty || state == EntryState.Added)
+                        {
+                            _transformTable.AddEntries(transformed, EntryState.Added);
+                        }
+                        else
+                        {
+                            _transformTable.ModifyEntriesFromPreviousTable(_previousTransformTable, transformed, _owner._comparer);
+                        }
+                    }
+                }
+            }
+
+            public void SetInputState(DriverStateTable.Builder builder)
+            {
+                builder.SetInputTable(_owner._filterNode, _filterTable.ToImmutableAndFree());
```

Comment:
Why is this `SetInputTable` and not `SetInputValue` given that `_filterNode` is an `InputNode<T>`?

Reply:
Addressed in follow up PR.

---
In reply to: [627695525](https://github.com/dotnet/roslyn/pull/53147#discussion_r627695525) [](ancestors = 627695525,627069285)